crypto/hpke/hpke_highlevel.c

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright 2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* An OpenSSL-based high level interface to HPKE */
+#include <stddef.h>
+#include <openssl/hpke.h>
+#include <openssl/core_names.h>
+#include <openssl/crypto.h>
+
+int OSSL_HPKE_BOX_seal(OSSL_LIB_CTX *libctx, const char *propq,
+                          const char *suitename,
+						  const unsigned char *receiver, size_t receiverlen,
+                          const unsigned char *pt, size_t ptlen,
+						  unsigned char **box, size_t *boxlen)
+{
+    OSSL_HPKE_SUITE suite;
+    size_t ctlen;
+    size_t enclen;
+    unsigned char *enc;
+    unsigned char *ct;
+    OSSL_HPKE_CTX *ctx;
+    unsigned char *out;
+    if (*box != NULL) {
+        return 0;
+    }
+    if (!OSSL_HPKE_str2suite(suitename, &suite)) {
+        return 0;
+    }
+    ctlen = OSSL_HPKE_get_ciphertext_size(suite, ptlen);
+    if (ctlen == 0) {
+        return 0;
+    }
+    enclen = OSSL_HPKE_get_public_encap_size(suite);
+    if (enclen == 0) {
+        return 0;
+    }
+    // Format: 4 byte length, encapsulation.
+    // 4 byte length, ciphertext.
+    // No trailing data allowed.
+    *boxlen = 8+enclen+ctlen;
+    *box = OPENSSL_malloc(*boxlen);
+    out = *box;
+    if (*box == NULL) {
+        *boxlen = 0;
+        return 0;
+    }
+    enc = out+4;
+    ctx = OSSL_HPKE_CTX_new(OSSL_HPKE_MODE_BASE, suite, OSSL_HPKE_ROLE_SENDER, libctx, propq);
+    if (ctx == NULL) {
+        OPENSSL_free(box);
+        *boxlen = 0;
+        return 0;
+    }
+    if (!OSSL_HPKE_encap(ctx, enc, &enclen, receiver, receiverlen, NULL, 0)) {
+        OSSL_HPKE_CTX_free(ctx);
+        OPENSSL_free(box);
+        *boxlen = 0;
+        return 0;
+    }
+    // It's possible that enclen is an overestimate and got changed.
+    out[0] = (unsigned char) (enclen);
+    out[1] = (unsigned char) (enclen>>8);
+    out[2] = (unsigned char) (enclen>>16);
+    out[3] = (unsigned char) (enclen>>24);
+    
+    ct = *box+8+enclen;
+    if (!OSSL_HPKE_seal(ctx, ct, &ctlen, NULL, 0, pt, ptlen)) {
+        OSSL_HPKE_CTX_free(ctx);
+        OPENSSL_free(*box);
+        *boxlen = 0;
+        return 0;
+    }
+    out[4+enclen] = (unsigned char) (ctlen);
+    out[4+enclen+1] = (unsigned char) (ctlen>>8);
+    out[4+enclen+2] = (unsigned char) (ctlen>>16);
+    out[4+enclen+3] = (unsigned char) (ctlen>>24);
```

Comment:
Have we invented a format for the "box" here? Or is this length prefixed approach in an RFC somewhere? If so can you provide a reference?

Reply:
We have invented a format.